uat/src/main/java/com.aws.greengrass/RegistrationSteps.java

```diff
@@ -0,0 +1,263 @@
+
+
+        // TODO: move this into iot steps.
+        IotThingSpec thingSpec = resources.create(IotThingSpec.builder()
+                .thingName(testContext.coreThingName())
+                .addThingGroups(IotThingGroupSpec.of(thingGroupName))
+                // Currently in case of hsm certificate is expected to be already created and in hsm.
+                .createCertificate(!testContext.hsmConfigured())
+                .certificateSpec(IotCertificateSpec.builder()
+                        .thingName(testContext.coreThingName())
+                        .csr(!csrPath.isEmpty() ? Files.readAllBytes(Paths.get(csrPath)).toString() : "")
+                        .existingArn(parameterValues.getString(FeatureParameters.EXISTING_DEVICE_CERTIFICATE_ARN)
+                                .orElse(""))
+                        .build())
+                .policySpec(resources.trackingSpecs(IotPolicySpec.class)
+                        .filter(p -> p.policyName().equals(testContext.testId().idFor("ggc-iot-policy")))
+                        .findFirst()
+                        .orElseGet(iotSteps::createDefaultPolicy))
+                .roleAliasSpec(IotRoleAliasSpec.builder()
+                        .name(testContext.testId().idFor("ggc-role-alias"))
+                        .iamRole(optionalIamRole.orElseGet(() ->
+                                resources.trackingSpecs(IamRoleSpec.class)
+                                        .filter(s -> s.roleName().equals(testContext.testId().idFor("ggc-role")))
+                                        .findFirst()
+                                        .orElseGet(iamSteps::createDefaultIamRole)
+                                        .resource()))
+                        .build())
+                .build());
+
+        try (InputStream input = getClass().getResourceAsStream(configFile)) {
+            setupConfig(
+                    thingSpec.resource(),
+                    thingSpec.roleAliasSpec(),
+                    IoUtils.toUtf8String(input),
+                    new HashMap<>());
+        }
+    }
+
+    private String getDefaultConfigName() {
+        if (testContext.hsmConfigured()) {
+            return DEFAULT_HSM_CONFIG;
+        }
+        return DEFAULT_CONFIG;
+    }
+
+    private void checkHSMConfigForPreInstalled() {
+        Path configPath = testContext.installRoot().resolve("config")
+                .resolve("effectiveConfig.yaml");
+        byte[] bytes = platform.files().readBytes(configPath);
+        try {
+            JsonNode config = mapper.readTree(bytes);
+            if (!config.get("services").hasNonNull("aws.greengrass.crypto.Pkcs11Provider")) {
+                throw new IOException("Pkcs11 is not properly configured on device");
+            }
+        } catch (IOException e) {
+            throw new ModuleProvisionException(e);
+        }
+    }
+
+    private void setupConfig(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method calls 33 other methods. By comparison, 99% of the methods in the CodeGuru reference dataset use fewer. This indicates the method is highly coupled with other methods. A method that is highly coupled with other methods is difficult to understand and its behavior might change unexpectedly when one of its referenced methods is updated. High coupling could also increase the integration test complexity, maintenance cost and technical debt.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 230-257 into a separate method.